hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/NonBlockingIOThread.java

```diff
@@ -307,6 +346,29 @@ public final void shutdown() {
         interrupt();
     }
 
+    // this method is always invoked in this thread
+    // after we have blocked for selector.select in #runSelectLoopWithSelectorFix
+    private void rebuildSelector() {
+        selectorRebuildCount.inc();
+        Selector newSelector = NonBlockingIOThread.newSelector(logger);
+        Selector oldSelector = this.selector;
+        this.selector = newSelector;
+
+        // cancel the old keys
```

Comment:
One thing just crossed my head. Probably you can rebuild the selection keys here as well because through the selectionkeys you can read all read/write handlers and then force a rebuild of the selection key. No need to rely on the onFailure.